File: examples/PHP/spworker.php
<?php
/*
 * Simple Pirate worker
 * Connects REQ socket to tcp://*:5556
 * Implements worker part of LRU queueing
 * 
 * @author Ian Barber <ian(dot)barber(at)gmail(dot)com>
 */
include "zmsg.php";

$context = new ZMQContext();
$worker = new ZMQSocket($context, ZMQ::SOCKET_REQ);

//  Set random identity to make tracing easier
$identity = sprintf ("%04X-%04X", rand(0, 0x10000), rand(0, 0x10000));
$worker->setSockOpt(ZMQ::SOCKOPT_IDENTITY, $identity);
$worker->connect("tcp://localhost:5556");

//  Tell queue we're ready for work
printf ("I: (%s) worker ready%s", $identity, PHP_EOL);
$worker->send("READY");

$cycles = 0;
while(true) {
	$zmsg = new Zmsg($worker);
	$zmsg->recv();
	$cycles++;
	
	//  Simulate various problems, after a few cycles
	if($cycles > 3 && rand(0, 3) == 0) {
		echo "I: (%s) simulating a crash", $identity, PHP_EOL;
		break;
	} else if($cycles > 3 && rand(0, 3) == 0) {
		echo "I: (%s) simulating CPU overload", $identity, PHP_EOL;
		sleep(5);
	}
	printf ("I: (%s) normal reply - %s%s", $identity, $zmsg->body(), PHP_EOL);
	sleep(1); // Do some heavy work
	$zmsg->send();
}

File: examples/C++/spworker.cpp
//
//  Simple Pirate worker
//  Connects REQ socket to tcp://*:5556
//  Implements worker part of LRU queueing
//
//  Andreas Hoelzlwimmer <andreas.hoelzlwimmer@fh-hagenberg.at>
#include "zmsg.hpp"

int main (void)
{
    srandom ((unsigned) time (NULL));

    zmq::context_t context(1);
    zmq::socket_t worker(context, ZMQ_REQ);

    //  Set random identity to make tracing easier
    std::string identity = s_set_id(worker);
    worker.connect("tcp://localhost:5556");

    //  Tell queue we're ready for work
    std::cout << "I: (" << identity << ") worker ready" << std::endl;
    s_send (worker, "READY");

    int cycles = 0;
    while (1) {
        zmsg zm (worker);

        //  Simulate various problems, after a few cycles
        cycles++;
        if (cycles > 3 && within (5) == 0) {
            std::cout << "I: (" << identity << ") simulating a crash" << std::endl;
            zm.clear ();
            break;
        }
        else
        if (cycles > 3 && within (5) == 0) {
            std::cout << "I: (" << identity << ") simulating CPU overload" << std::endl;
            sleep (5);
        }
        std::cout << "I: (" << identity << ") normal reply - " << zm.body () << std::endl;
        sleep (1);              //  Do some heavy work
        zm.send(worker);
    }
    return 0;
}
